### ISSUE TITLE
Telepot does not wait for Twitter API to return data

### ISSUE THREAD

**Issue by albertsy2** [nickoala/telepot#318]
Hi!

I have a Telegram bot (using the telepot wrapper) which I programmed in Python. When I send a specific command to the bot by my phone, it goes to the internet using API's and returns data to my phone. Depending on the command I send it, the bot connects to OpenWeatherMap (grabs my local weather info) or connects to Twitter (searches for one particular keyword that I need to keep an eye on).

- When I run the Telegram python in Pixel OS (in IDLE), it works OK (I can get data from both bots).
- When I run the Telegram python in the Command Line Interface, it also works OK (I can get data from both bots).
- But when I autorun the Telegram python on startup (through a systemd service or even cron), the Twitter API does not return anything. The OpenWeatherMap API runs OK. (So it's not a problem with my modem or router.)

I am wondering if the Telegram bot runs too fast in the background and does not wait for the Twitter API to return data.
(I even changed my module from twython to tweepy and it still behaves the same way [It does not return Twitter data when auto-run on startup]. So it's not the wrapper. )

--- 

I have been able to lightswitch the issue in Pixel OS IDLE environment. Here is my snippet of code. I can lightswitch the issue by putting a sleep (5)  in the try statement. (The code works in IDLE if the sleep is not there. I even tried going down to sleep (0.0000000000001) and it still does not wait; it returns the exception.) This shows that telepot is really not waiting for twitter API to return data and [maybe] demands an instantaneous reply(?). Is there some way to make telepot wait? Thanks! I am running the code in Python 2.7.9 on a Raspberry Pi 3 running Pixel. 

    elif command == '/smrt_tweets':   
        try:
            sleep (5) 
            SMRT_tweets_str = tweepy_search_module.tweepy_get_tweets(10)            	                        
        except:	    
            SMRT_tweets_str = 'Telegram connection issue to Twitter API'

**Comment by das7pad** [comment]
The sleep call seems to be nonsense here. It just delays further function calls.

Could you change the `try-except`-block to print the Exception?
```python
try:
   ...
except Exception as err:
    print(repr(err))
```

**Comment by albertsy2** [comment]
try:
            sleep (5)
            print 'looking for tweets!'
            SMRT_tweets_str = tweepy_search_module.tweepy_get_tweets(10)

        except telepot.exception.TelegramError:
            SMRT_tweets_str = '%d: %s' % (telepot.exception.TelegramError.error_code, telepot.exception.TelegramError.description)
        
        except:	    
            e = sys.exc_info()[0]
            SMRT_tweets_str = '%s' % e
            
        print len(SMRT_tweets_str)       
        print (chat_id)
        print (SMRT_tweets_str)
        bot.sendMessage(chat_id,SMRT_tweets_str)      


This code returns: 
_<type 'exceptions.NameError'>_

It doesn't look like a Telepot or Telegram error. 

If I remove the sleep, it returns the normal Twitter API data that I expect. Thanks!

**Comment by albertsy2** [comment]
Going deeper into NameError, I added this except block...

        except NameError as e: 	    
            SMRT_tweets_str = e.message.split("'")[1]

Then this returns:

<type 'exceptions.UnicodeEncodeError'>

**Comment by das7pad** [comment]
The first except block is not reachable, as there is no telepot call.

Please use the except block I posted above to print the `repr`esentation of the exception (which includes it's type and message).

**Comment by albertsy2** [comment]
Hi! Here is the result of... 

        except Exception as err:
            print(repr(err))

When autorun by cron on bootup:

UnicodeEncodeError('ascii', u'Twitter>>\nSun Oct 22 20:02:54 +0800 2017 @AWCavs : At @smrt_singapore \u2014 https://t.co/mfEGCpBysA\n', 70, 71, 'ordinal not in range(128)')

When running in Pixel  IDLE, the Shell window shows:

Sun Oct 22 20:02:54 +0800 2017 @AWCavs : At @smrt_singapore — https://t.co/mfEGCpBysA

In Telegram on my phone, I also see: 

Sun Oct 22 20:02:54 +0800 2017 @AWCavs : At @smrt_singapore — https://t.co/mfEGCpBysA

... and the message shows a preview of that URL. Wonder if that object is the "ordinal not in range".

**Comment by das7pad** [comment]
`—` aka `\u2014` seems to be not handled properly by Python2.x without additional conversation.

Anyways - great to hear that the tweets made it to your Telegram chat!

**Comment by albertsy2** [comment]
Wait, wait...

The problem was that my Telegram fails when I autorun on startup, using cron or systemd. But the Telegram is OK when running in Pixel IDLE.

But thanks to your replies and feedback, I was able to track the exception as being an encoding problem. Apparently, the encoding used when Python is autorun is different from when it is run in IDLE! Beats me why!

I was able to fix the program by adding this to the beginning of my code:

```
import sys  
reload(sys)  
sys.setdefaultencoding('utf8')
```

Thanks! I suppose this can be closed, but could still be a bug...
